pkg/block/fetcher.go

```diff
@@ -286,12 +379,30 @@ type response struct {
 	incompleteView bool
 }
 
-func (f *BaseFetcher) fetchMetadata(ctx context.Context) (interface{}, error) {
+func (r response) metasCopy() map[ulid.ULID]*metadata.Meta {
+	metas := make(map[ulid.ULID]*metadata.Meta, len(r.metas))
+	for id, m := range r.metas {
+		metas[id] = m
+	}
+	return metas
+}
+
+// deletionMarksCopyForFilter makes a copy of deletion marks map, suitable for passing to Filter method.
+func (r response) deletionMarksCopyForFilter() map[ulid.ULID]*metadata.DeletionMark {
+	marks := make(map[ulid.ULID]*metadata.DeletionMark, len(r.marks))
+	for id, m := range r.marks {
+		marks[id] = &m.mark
```

Comment:
The reason why I commented about storing by copy/reference in `cachedDeletionMark` is due to the factor that here we work with pointers, while in the cache we store by copy. It just looked weird to me, but I'm not feeling strong about this.